Add event-notification feature file template

bigludo7 opened the pull request:
#### What type of PR is this?

<!-- Add one of the following kinds: -->

* documentation


#### What this PR does / why we need it:

This is a proposal of Subscription feature template.
I used the nice work done by @maxl2287 in several subscriptions API ;)


#### Which issue(s) this PR fixes:

<!-- Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`. -->

Fixes #293 


#### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If indicated yes above, please describe the breaking change(s). -->

#### Special notes for reviewers:



#### Changelog input

```
 release-note
- add a template for subscription test case.

```

#### Additional documentation 

This section can be blank.



```
docs

```

patrice-conil commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+    When the request "createSubscription" is sent 
+    Then the response code is 201
+    And the response header "Content-Type" is "application/json"
+    And the response header "x-correlator" has same value as the request header "x-correlator"
```
Maybe we have to mention Location header that points to created subscription?

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+
+  @<xxx>_subscriptions_32_get_unknown_<xxx>_subscription_for_a_device
+  Scenario:  Get method for <xxx> subscription with subscription-id unknown to the system  
+    Given the path parameter "subscriptionId" is set to the value unknown to system
```
```suggestion
    Given the path parameter "subscriptionId" is set to a value not corresponding to any existing subscription

```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+
+  @<xxx>_subscriptions_36_delete_invalid_<xxx>_subscription
+  Scenario:  Delete <xxx> subscription with subscription-id unknown to the system
+    Given the path parameter "subscriptionId" is set to the value unknown to system
```
```suggestion
    Given the path parameter "subscriptionId" is set to a value not corresponding to any existing subscription

```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+    And the response property "$.code" is "NOT_FOUND"
+    And the response property "$.message" contains a user friendly text
+
+########Specific Subscription error scenarios when Device object is  used ################
```
As we have these scenarios included in dedicated artifacts we may just reference the other artifacts to not maintain equivalent scenarios in several places

bigludo7 replied:
Yes agree with your proposal @jlurien  to remove them to avoid redundancy.
Before to remove them looking for OK from other "usual suspects": @PedroDiez @rartych @patrice-conil  OK for you?

jlurien replied:
Yes, I have updated #479, so now artifact is compatible with APIs having device or phoneNumber in paths other than $.xxx

jlurien reviewed (commented):
We may include some additional scenario to test the "initialEvent". e.g.

````
 @<xxx>_subscriptions_XX_subscription_creation_initial_event
  Scenario: Receive initial event notification on creation
    Given the API supports initial events to be sent
    And a valid subscription request body with property "$.config.initialEvent" set to true
    When the request "create<xxx>Subscription" is sent
    Then the response code is 201 or 202	
    And an event notification of the subscribed type is received on callback-url
    And notification body complies with the OAS schema at "#/components/schemas/Subscription"
```
This optional scenario may apply for some implementations
````
 @<xxx>_subscriptions_XX_subscription_creation_multievent_not_supported
  Scenario: Multi event subscription not supported
    Given the API provider only allows one event to be subscribed per subscription request
    And a valid <xxx> subscription request body 
    And the request body property "$.types" is set to an array of 2 valid items
    When the request "create<xxx>Subscription" is sent
    Then the response code is 422
    And the response property "$.status" is 422
    And the response property "$.code" is "MULTIEVENT_SUBSCRIPTION_NOT_SUPPORTED"
    And the response property "$.message" contains a user friendly text
```

jlurien commented:
@bigludo7 please ask me to re-review when you think is ready

bigludo7 commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+########Specific Subscription error scenarios when Device object is  used ################
+### These scenarios must be present if the device object is used in the subscription  ####
+
+  @<xxx>_subscriptions_40_create_with_identifier_mismatch
```
Yes correct - removed.

bigludo7 commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+    Then the subscription is expired
+    Then event notification "subscription-ended" is received on callback-url
+    And notification body complies with the OAS schema at "##/components/schemas/EventSubscriptionEnds"
+    And type="org.camaraproject.<xxx>-subscriptions.v0.subscription-ended"
```
fixed.

bigludo7 commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+	
+  @<xxx>_subscriptions_08_subscription_ends_on_expiry
+  Scenario: Receive notification for subscription-ended event on expiry
+    Given a valid <xxx> subscription request body 
```
Thanks - replaced with your proposal

bigludo7 commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+    And the request property "$.sinkCredential.accessToken" is valued with a valid value
+    And the request property "$.sinkCredential.accessTokenExpiresUtc" is valued with a valid value
+    Then the response property "$.status" is 400
+    And the response property "$.code" is "INVALID_TOKEN" or "INVALID_ARGUMENT"
```
Yep - removed !

bigludo7 commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,356 @@
+    And header "Authorization" set to an invalid access token
+    When the request "createSubscription" is sent
+    Then the response status code is 401
+    And the response header "Content-Type" is "application/json"
```
Correct - removed.

bigludo7 commented:
@rartych @jlurien @PedroDiez @patrice-conil Thanks all for your review !
I've considered all of them except 2 that need additional work

- the one from my colleague @patrice-conil about the header location (will check with Patrice how to accommodate this one)
- the comment about keeping the `device` error cases - I tend to think we can keep it for now waiting that we have a specific .feature file for `device` error.

rartych commented:
Let's keep the guideline as normative only source of  truth - starting new issue is a way to go.

bigludo7 commented:
@rartych @jlurien @PedroDiez Thanks for your feedback.

- `device` related TC removed (and added a note in introduction part)
- No mention to the header `location`

PedroDiez commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  Background: Common <xxx> Subscriptions setup
+    Given the resource "{apiroot}/<xxx>-subscriptions/vwip/" as <xxx> base-url
+    And the header "Authorization" is set to a valid access token
+    And the header "x-correlator" follows the pattern ^[a-zA-Z0-9-_:;.\/<>{}]{0,256}$
```
```suggestion
    And the header "x-correlator" complies with the schema at "#/components/schemas/XCorrelator"

```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+    Then the response code is 201
+    And the response header "Content-Type" is "application/json"
+    And the response header "x-correlator" has the same value as the request header "x-correlator"
+    And the response body complies with the OAS schema at "/components/schemas/Subscription"
```
```suggestion
    And the response body complies with the OAS schema at "#/components/schemas/Subscription"
```

We should then use `#/components/schemas/XXX` in all references

bigludo7 replied:
fixed

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  Scenario:  Create <xxx> subscription with invalid parameter
+    Given the request body is not compliant with the schema "/components/schemas/SubscriptionRequest"
+    When the request "create<xxx>Subscription" is sent 
+
```
```suggestion
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+
+  @<xxx>_subscriptions_20_create_<xxx>_subscription_with_invalid_parameter
+  Scenario:  Create <xxx> subscription with invalid parameter
+    Given the request body is not compliant with the schema "/components/schemas/SubscriptionRequest"
```
```suggestion
    Given the request body is not compliant with the schema "#/components/schemas/SubscriptionRequest"
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  Scenario: Expiry time in past
+    Given a valid <xxx> subscription request body 
+    And request body property "$.config.subscriptionExpireTime" in the past
+
```
```suggestion
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  Scenario: Subscription creation with invalid protocol
+    Given a valid <xxx> subscription request body 
+    When the request "create<xxx>Subscription" is sent
+    And the request property "$.protocol" is not set to "HTTP"
```
When step after all request setup steps

bigludo7 replied:
fixed

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscription_24_invalid_credential
+  Scenario: Subscription creation with invalid credential
+    Given a valid <xxx> subscription request body 
+    When the request "create<xxx>Subscription" is sent
```
idem

bigludo7 replied:
fixed

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscription_25_invalid_token
+  Scenario: Subscription creation with invalid token
+    Given a valid <xxx> subscription request body
+    When the request "create<xxx>Subscription" is sent
```
idem

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscription_33_invalid_url
+  Scenario: Subscription creation with invalid url
+    Given a valid <xxx> subscription request body 
+    When the request "create<xxx>Subscription" is sent
```
When step just before Then steps

bigludo7 replied:
fixed

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscriptions_36_delete_invalid_<xxx>_subscription
+  Scenario:  Delete <xxx> subscription with subscription-id unknown to the system
+    Given the path parameter "subscriptionId" is set to a value not corresponding to any existing subscription
+
```
```suggestion
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscriptions_30_expired_access_token_for_get_subscription
+  Scenario: Expired access token for get subscription
+    Given the header "Authorization" is set to expired token
+    When the request "retrieve<xxx>Subscription" is sent
```
Missing step: And path parameter "subscriptionId" is set to the identifier of an existing <xxx> subscription

bigludo7 replied:
fixed

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscriptions_31_invalid_access_token_for_get_subscription
+  Scenario: Invalid access token for get subscription
+    Given the header "Authorization" set to an invalid access token
+    When the request "retrieve<xxx>Subscription" is sent
```
Missing step: And path parameter "subscriptionId" is set to the identifier of an existing subscription

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscriptions_33_no_authorization_header_for_delete_subscription
+  Scenario: No Authorization header for delete subscription
+    Given header "Authorization" is set without a token
+    When the request "delete<xxx>Subscription" is sent 
```
Missing step: And path parameter "subscriptionId" is set to the identifier of an existing subscription

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscriptions_34_expired_access_token_for_delete_subscription
+  Scenario: Expired access token for delete subscription
+    Given header "Authorization" is set with an expired token
+    When the request "delete<xxx>Subscription" is sent
```
Missing step: And path parameter "subscriptionId" is set to the identifier of an existing subscription

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscriptions_35_invalid_access_token_for_delete_subscription
+  Scenario: Invalid access token for delete subscription
+    Given header "Authorization" set to an invalid access token
+    When the request "delete<xxx>Subscription" is sent
```
Missing step: And path parameter "subscriptionId" is set to the identifier of an existing subscription

jlurien reviewed (dismissed):
LGTM, with minor suggestion

PedroDiez commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,369 @@
+
+########################### Subscription list retrieval scenarios #####################################
+
+  @<xxx>_subscriptions_60_no_authorization_header_for_list_subscription
```
align test numbering

- create
- get
- list
- delete

and final multievent case

bigludo7 replied:
Done !

PedroDiez reviewed (commented):
Some editorial comments and suggest test number alignment since test 20

bigludo7 commented:
Thanks @PedroDiez & @jlurien 
I've fixed the file by removing these empty lines that are added when I copy the file from VS :(
Pedro, I've also fixed the TC numbering (and move the one related to the url).

PedroDiez reviewed (approved):
LGTM

rartych reviewed (dismissed):
LGTM

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  # We designed as <xxx> the subscription resource prefix.
+  # We designed as <x> the event version. 
+  #
+  # If the subscription leverages 'device' object the following indication must be present:
```
```suggestion
  # If the subscription leverages the 'device' object the following indication must be present:
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  #
+  # If the subscription leverages 'device' object the following indication must be present:
+  #    Implementation indications:
+  #      * List of device identifier types which are not supported, among: phoneNumber, networkAccessIdentifier, ipv4Address, ipv6Address
```
```suggestion
  #      * List of device identifier types which are not supported, such as: phoneNumber, networkAccessIdentifier, ipv4Address, ipv6Address
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  #        A sink-url identified as "callbackUrl", which receives notifications
+  #        + Add here the specific testing asset(s) required to test the API
+  #
+  # References to OAS spec schemas refer to schemas specifies in <xxx>-subscriptions.yaml
```
```suggestion
  # References to OAS spec schemas refer to schemas specified in <xxx>-subscriptions.yaml
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  # References to schemas starting with the # symbol are JSON Pointers from the root of the OAS document: <xxx>-subscriptions.yaml, Schema names are aligned with the event-subscription-template.yaml artifact.
+  #
+  # IMPORTANT: 
+  # 1/ This file must be completed with test cases specific to the subscription type managed by the API.
```
```suggestion
  # 1/ This file must be completed with the test cases specific to the subscription type managed by the API.
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  #
+  # IMPORTANT: 
+  # 1/ This file must be completed with test cases specific to the subscription type managed by the API.
+  # 2/ Specific Subscription error scenarios when Device object is used must be added. These scenarios are available in CAMARA Github.   ################
```
```suggestion
  # 2/ Specific Subscription error scenarios when the Device object is used must be added. These scenarios are available in CAMARA Github.   ################
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+# Note: Depending on the API managed personal data specific scenario update may be require to specify use of 2-legs or 3-legs access token.
+
+  @<xxx>_subscriptions_01_Create_<xxx>_subscription_sync
+  Scenario:  Create <xxx> subscription (sync creation)
```
```suggestion
  Scenario: Create <xxx> subscription (sync creation)
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  @<xxx>_subscription_22_creation_with_invalid_eventType
+  Scenario: Subscription creation with invalid event type
+    Given a valid <xxx> subscription request body 
+    And the request body property "$.types" is set to invalid value  
```
```suggestion
    And the request body property "$.types" is set to an invalid value  
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  Scenario: Subscription creation with invalid url
+    Given a valid <xxx> subscription request body
+    And the request property "$.protocol" is set to "HTTP"
+    And the request property "$.sink" is set to "azerty"    
```
```suggestion
    And the request property "$.sink" is set to "some_foo_value"    
```

jlurien commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,366 @@
+  Scenario: Multi event subscription not supported
+    Given the API provider only allows one event to be subscribed per subscription request
+    And a valid subscription request body
+    And the request body property "$.types" is set to an array of 2 valid items
```
```suggestion
    And the request body property "$.types" is set to an array with 2 valid items
```

jlurien reviewed (commented):
Some suggestions (by Copilot mostly :D)

rartych reviewed (dismissed):
Reapproved

PedroDiez commented on artifacts/camara-cloudevents/event-subscription-template.feature:
```diff
@@ -0,0 +1,369 @@
+Feature: Camara Template Subscriptions API, v{version here} - Operations on subscriptions
+
+  # This feature file is to be used by CAMARA subproject when an event subscription resource is provided.
+
```
Remove Lines 5 and 7 (empty lines)
LGTM in advance

PedroDiez replied:
Done

jlurien reviewed (dismissed):
LGTM

PedroDiez reviewed (approved):
LGTM
